internal/middleware/compression.go

```diff
@@ -15,12 +15,60 @@
 package middleware
 
 import (
+	"bytes"
 	"compress/gzip"
 	"compress/zlib"
 	"io"
 	"net/http"
 )
 
+const (
+	headerContentEncoding = "Content-Encoding"
+	headerValueGZIP       = "gzip"
+)
+
+type CompressRoundTripper struct {
+	http.RoundTripper
+	gzipWriter *gzip.Writer
+}
+
+func NewCompressRoundTripper(rt http.RoundTripper) *CompressRoundTripper {
+	return &CompressRoundTripper{
+		RoundTripper: rt,
+		gzipWriter:   gzip.NewWriter(nil),
+	}
+}
+
+func (r *CompressRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+
+	// Gzip the body.
+	buf := bytes.NewBuffer([]byte{})
+	r.gzipWriter.Reset(buf)
+	_, err := io.Copy(r.gzipWriter, req.Body)
+
+	defer req.Body.Close()
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := r.gzipWriter.Close(); err != nil {
```

Comment:
I am confused why you close this but the received one was not closed.

Reply:
Do you mean the response? The response is returned to the caller, they are responsible for closing.

Reply:
I mean the 4 lines above. Body doesn't seem to be closed when you receive it. You either do this and not close the body or close the body but not this.

Reply:
Fixed.